controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,1008 @@
+    bit = 0;
+    break;
+  case TIMER3_BASE:
+    ndx = 4;
+    bit = 1;
+    break;
+  case TIMER6_BASE:
+    ndx = 4;
+    bit = 4;
+    break;
+  case TIMER7_BASE:
+    ndx = 4;
+    bit = 7;
+    break;
+  case SPI2_BASE:
+    ndx = 4;
+    bit = 14;
+    break;
+  case SPI3_BASE:
+    ndx = 4;
+    bit = 15;
+    break;
+  case UART2_BASE:
+    ndx = 4;
+    bit = 17;
+    break;
+  case UART3_BASE:
+    ndx = 4;
+    bit = 18;
+    break;
+  case UART4_BASE:
+    ndx = 4;
+    bit = 19;
+    break;
+  case I2C1_BASE:
+    ndx = 4;
+    bit = 21;
+    break;
+  case I2C2_BASE:
+    ndx = 4;
+    bit = 22;
+    break;
+  case I2C3_BASE:
+    ndx = 4;
+    bit = 23;
+    break;
+
+  case I2C4_BASE:
+    ndx = 5;
+    bit = 1;
+    break;
+
+  case TIMER1_BASE:
+    ndx = 6;
+    bit = 11;
+    break;
+  case SPI1_BASE:
+    ndx = 6;
+    bit = 12;
+    break;
+  case UART1_BASE:
+    ndx = 6;
+    bit = 13;
+    break;
+  case TIMER15_BASE:
+    ndx = 6;
+    bit = 16;
+    break;
+  case TIMER16_BASE:
+    ndx = 6;
+    bit = 17;
+    break;
+  }
+
+  // If the input address wasn't found then it's definitly
+  // a bug.  I'll just loop forever here causing the code
+  // to crash.  That should make it easier to find the
+  // bug during development.
+  if (ndx < 0) {
+    IntDisable();
+    while (1) {
+    }
+  }
+
+   // Enable the clock of the requested peripherial
+   RCC_Regs *rcc = reinterpret_cast<RCC_Regs *>(RCC_BASE);
+   rcc->periphClkEna[ndx] |= (1<<bit);
+}
+
+/******************************************************************
+ * Interrupt vector table.  The interrupt vector table is a list of
+ * pointers to the various interrupt functions.  It is stored at the
+ * very start of the flash memory.
+ *****************************************************************/
+
+// Fault handlers
+static void fault()
+{
+   while(1){}
+}
+
+static void NMI()           { fault(); }
+static void FaultISR()      { fault(); }
+static void MPUFaultISR()   { fault(); }
+static void BusFaultISR()   { fault(); }
+static void UsageFaultISR() { fault(); }
+static void BadISR()        { fault(); }
+
+extern "C" void Reset_Handler();
+__attribute__((section(".isr_vector"))) void (*const vectors[])() = {
+    // The first entry of the ISR holds the initial value of the
+    // stack pointer.  The ARM processor initializes the stack
+    // pointer based on this address.
+    reinterpret_cast<void (*)()>(&systemStack[SYSTEM_STACK_SIZE]),
+
+    // The second ISR entry is the reset vector which is an
+    // assembly language routine that does some basic memory
+    // initilization and then calls main().
+    // Note that the LSB of the reset vector needs to be set
+    // (hence the +1 below).  This tells the ARM that this is
+    // thumb code.  The cortex m4 processor only supports
+    // thumb code, so this will always be set or we'll get
+    // a hard fault.
+    reinterpret_cast<void (*)()>(reinterpret_cast<uint32_t>(Reset_Handler) + 1),
```

Comment:
uintptr_t?

Reply:
Not quite sure what you mean here.

Reply:
OK, never mind.  I see what you mean, that does work better.
I made that change but it pushed me over 80 characters on that line so that clang-format script broke it into two.  Very annoying and ugly, I'm not a fan of what that script does to my beautiful code!